src/ext/default.c

```diff
@@ -26,41 +27,61 @@
 // normalize TF by number of tokens (weighted)
 #define NORM_DOCLEN 2
 
-// recursively calculate tf-idf
-static double tfidfRecursive(const RSIndexResult *r, const RSDocumentMetadata *dmd) {
+#define EXPLAIN(indexResult, fmt, args...)                      \
+        if (explain) {                                           \
+          asprintf((char ** restrict)&indexResult->scoreExplainStr, fmt, ##args);      \
+        } 
 
+// recursively calculate tf-idf
+static double tfidfRecursive(const RSIndexResult *r, const RSDocumentMetadata *dmd, bool explain) {
   if (r->type == RSResultType_Term) {
-    return r->weight * ((double)r->freq) * (r->term.term ? r->term.term->idf : 0);
+    double idf = r->term.term ? r->term.term->idf : 0;
+    double res = r->weight * ((double)r->freq) * idf;
+    EXPLAIN(r, "TFIDF %.2f = Weight %.2f * TF %d * IDF %.2f", res, r->weight, r->freq, idf);
+    return res;
   }
   if (r->type & (RSResultType_Intersection | RSResultType_Union)) {
     double ret = 0;
     for (int i = 0; i < r->agg.numChildren; i++) {
-      ret += tfidfRecursive(r->agg.children[i], dmd);
+      ret += tfidfRecursive(r->agg.children[i], dmd, explain);
     }
+    EXPLAIN(r, "%.2f = Weight %.2f * children TFIDF %.2f",  r->weight * ret, r->weight, ret)
     return r->weight * ret;
   }
+  EXPLAIN(r, "TFIDF %.2f = Weight %.2f * Frequency %d", r->weight * (double)r->freq, r->weight, r->freq);
   return r->weight * (double)r->freq;
 }
 
 /* internal common tf-idf function, where just the normalization method changes */
 static inline double tfIdfInternal(const ScoringFunctionArgs *ctx, const RSIndexResult *h,
                                    const RSDocumentMetadata *dmd, double minScore, int normMode) {
-  if (dmd->score == 0) return 0;
-  double norm = normMode == NORM_MAXFREQ ? (double)dmd->maxFreq : dmd->len;
+  bool explain = ctx->scoreflags & SCORE_F_WITH_EXPLANATION;
+  if (dmd->score == 0) {
+    EXPLAIN(h, "Document score is 0");
+    return 0;
+  }
+  uint32_t norm = normMode == NORM_MAXFREQ ? dmd->maxFreq : dmd->len;
+  double rawTfidf = tfidfRecursive(h, dmd, explain);
+  double tfidf = dmd->score * rawTfidf / norm;
 
-  double tfidf = dmd->score * tfidfRecursive(h, dmd) / norm;
   // no need to factor the distance if tfidf is already below minimal score
   if (tfidf < minScore) {
+    EXPLAIN(h, "TFIDF score of %.2f is smaller than minimum score %.2f", tfidf, minScore);
     return 0;
   }
 
-  tfidf /= (double)ctx->GetSlop(h);
+  int slop = ctx->GetSlop(h);
+  tfidf /= slop;
+  char *tempStr = h->scoreExplainStr;
+  EXPLAIN(h, "%s. Final TFIDF : %.2f * document score %.2f / norm %d / slop %d",
+          h->scoreExplainStr, rawTfidf, dmd->score, norm, slop);
+  if (h->scoreExplainStr != tempStr) free(tempStr);
```

Comment:
I think the if is not needed, its ok to just free, also use rm_free and rm_alloc everywhere.